rewrite-gradle/src/main/java/org/openrewrite/gradle/MigrateDependenciesToVersionCatalog.java

```diff
@@ -0,0 +1,1164 @@
+import org.openrewrite.kotlin.tree.K;
+import org.openrewrite.marker.Markers;
+import org.openrewrite.marker.SearchResult;
+import org.openrewrite.properties.PropertiesVisitor;
+import org.openrewrite.properties.tree.Properties;
+import org.openrewrite.toml.TomlParser;
+import org.openrewrite.toml.tree.Toml;
+
+import java.nio.file.Paths;
+import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static java.util.Collections.*;
+
+@Value
+@EqualsAndHashCode(callSuper = false)
+public class MigrateDependenciesToVersionCatalog extends ScanningRecipe<MigrateDependenciesToVersionCatalog.DependencyAccumulator> {
+
+    private static final MethodMatcher DEPENDENCY_DSL_MATCHER = new MethodMatcher("DependencyHandlerSpec *(..)");
+    private static final MethodMatcher PROJECT_MATCHER = new MethodMatcher("DependencyHandlerSpec project(..)");
+    private static final Pattern DEPENDENCY_STRING_PATTERN = Pattern.compile("([^:]+):([^:]+):([^:@]+)(@.+)?");
+    private static final Pattern DEPENDENCY_MAP_PATTERN = Pattern.compile("group:\\s*['\"]([^'\"]+)['\"],\\s*name:\\s*['\"]([^'\"]+)['\"],\\s*version:\\s*['\"]([^'\"]+)['\"]");
+    private static final String CATALOG_PATH = "gradle/libs.versions.toml";
+    private static final String GRADLE_PROPERTIES = "gradle.properties";
+
+    @Override
+    public String getDisplayName() {
+        return "Migrate Gradle project dependencies to version catalog";
+    }
+
+    @Override
+    public String getDescription() {
+        //language=markdown
+        return "Migrates Gradle project dependencies to use the [version catalog](https://docs.gradle.org/current/userguide/platforms.html) feature. " +
+                "Supports migrating dependency declarations of various forms:\n" +
+                " * `String` notation: `\"group:artifact:version\"`\n" +
+                " * `Map` notation: `group: 'group', name: 'artifact', version: 'version'`\n" +
+                " * Property references: `\"group:artifact:$version\"` or `\"group:artifact:${version}\"`\n\n" +
+                "The recipe will:\n" +
+                " * Create a `gradle/libs.versions.toml` file with version declarations\n" +
+                " * Replace dependency declarations with catalog references (e.g., `libs.springCore`)\n" +
+                " * Migrate version properties from `gradle.properties` to the version catalog\n" +
+                " * Preserve project dependencies unchanged\n\n" +
+                "**Note:** If a version catalog already exists, the recipe will not modify it.";
+    }
+
+    static class DependencyAccumulator {
+        final Map<String, DependencyInfo> dependencies = synchronizedMap(new LinkedHashMap<>());
+        final Set<String> configurations = synchronizedSet(new LinkedHashSet<>());
+        final Set<String> propertyNamesToRemove = synchronizedSet(new LinkedHashSet<>());
+        final Map<String, String> propertyValues = synchronizedMap(new LinkedHashMap<>());
```

Comment:
This synchronization shouldn't be necessary, recipe execution is single threaded.

Reply:
done